### ISSUE TITLE
mmapstorage: lock mapped files

### ISSUE THREAD

**Issue by mpranj** [ElektraInitiative/libelektra#2248]
## Problem

When mmapstorage returns from the `plugin->kdbGet()` function, the returned keyset is inside a mmap region backed by the configuration/mmap file. There is no specified behaviour, if the file changes after the `mmap()` call. On my system, the changes become visible in the mapped region. Overwriting a currently mapped file therefore destroys old mappings. The previously returned keysets (in elektrified client programs) are therefore invalid and result in SIGBUS/SIGSEGV/unspecified behaviour.

The problem is partly mitigated by mmapstorage remembering current mappings inside the `plugin->data`. Since the data is not shared, this does not work for other plugin instances or other KDB instances.

## Considered Solutions

1. Use `fcntl()` to set advisory locks, so at least two instances of KDB can not overwrite the underlying file while it is mapped. This is already implemented in #2246, but that makes overwriting a mapped config/cache file impossible.
2.  Use a different type of communication to synchronize write access to the mmap file. (shared memory,...?)
3. Use mmap() (or fread/fwrite) only for read/write operations and immediately close the mapping. Return a "normal" malloc()-ed keyset. A quick test suggests that this would be merely 30x faster than dump (instead of the current 67x improvement). I would have to implement it properly to be sure though.

## Implications 

What is more of interest for the team is probably how this effects the global cache.

One possibility is that we could keep the mapped keyset only internally, and expose only malloc()-ed keysets to the user. The downside is that we run into the same problems when a KDB instance wants to overwrite a cache file, which might still be mapped by another KDB instance. (-> we then need a more complex solution)

The more I think about it, the more I lean towards the version (3) where we use mmap() only to read and write.

Any ideas by the team are welcome. @markus2330 you might also want to comment.

**Comment by mpranj** [comment]
After spending some more time on solution 3:

Pros:

- If done properly, it seems to be as fast or even **faster** than my initial mmapstorage proposal.
- It is rather simple, much simpler than approach 2.
- It does not leave open file descriptors or mmap() regions (which is the case in approach 1 and 2).
- The mmap cache files can be updated/overwritten easily, since the plugin leaves no open mappings.
- It resolves the problem above completely.

Cons:

- More memory consumption. (debatable, depends on the data usage patterns and mmap implementation)
- Performance relies on reading data into a large block allocated with malloc(). (still consumes the same amount of memory as `dump` + constant overhead)

**Comment by markus2330** [comment]
Isn't this the same bug we already had? If you unlink the file before writing to it, clients should still see the old content.

Or do you mean the problem is that kdbGet() does not work if the mmaped file is currently in the process of being written? If the file is not yet fully written, you can either wait (locking) or simply not use the cache. (If the file is currently being written should be detected easily because you know the length of the file?)

**Comment by mpranj** [comment]
The problem is on a completely different level, but it is the same problem, yes. Since Elektra runs in the address space of other elektrified programs, there is no sharing of the list of currently linked files. This works only inside a single instance.

Let's take a look at the following scenario:

1. We have two elektrified programs A and B running.
2. Program A successfully opens a KDB handle + cache and leaves the mapping open (since that is what mmapstorage does).
3. Program B opens a KDB handle and gets the cache too, but it can overwrite the cache file. It does not know that program A has it currently mapped.

The list of mapped (linked) files is inside the plugin->data, which is not shared across different processes. Therefore we need either some locking, or just create a copy of the data.
I have implemented both approaches. Locking #2246 is error prone, leads to further issues that need solving and is also slower. Instead of leaving open mappings, we can also create an in-memory copy of the data. My tests in #2258 show that this is less error prone and also **faster** in benchmarks.

**Comment by markus2330** [comment]
> address space of other elektrified programs
> only inside a single instance

Please use the terms "process" and "threads". I still do not understand which case is the problematic one. (Are A and B processes?)

> no sharing of the list of currently linked files

Why do we need that?

> but it can overwrite the cache file

Why would it override the cache file? It should unlink it before writing to it.

Btw. actually it should use rename so that everything is atomic and works without locking. For the storage plugin this is automatically done but for the global cache you need to implement this yourself.

> Instead of leaving open mappings, we can also create an in-memory copy of the data.

Does this really resolve the issue? One could already overwrite the file while the copy is done.

**Comment by mpranj** [comment]
> (Are A and B processes?)

Yes, A and B are distinct processes. I have actually [implemented a test](https://github.com/ElektraInitiative/libelektra/pull/2246/files#diff-dcf019647cad4930fd5271be9f829267R792) that shows exactly where this fails.

> Why do we need that?

Process A does not know which files are currently linked by process B. Process A also has no way of telling process B to "unlink" some file.

> Does this really resolve the issue? One could already overwrite the file while the copy is done.

It solves the issue for `mmapstorage`, since resolver locks the file if locking is enabled.

**Comment by markus2330** [comment]
> Process A also has no way of telling process B to "unlink" some file.

Communication is not needed. B (and A) always unlinks before writing to a file.

> It solves the issue for mmapstorage, since resolver locks the file if locking is enabled.

Yes, the resolver already does the unlinking via rename.

**Comment by mpranj** [comment]
Thank you for reiterating this!

I completely misunderstood what you meant this whole time. I implemented a "custom unlinking" approach by copying a keyset out of the mapped region, once the file has to be overwritten. (this is where we would need inter-process communication)
I overlooked that this is naturally implemented by `unlink()` in POSIX. 😢 

The issue will be resolved when #2262 is ready to merge.
